# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [2]:
# Run this code:
salaries = pd.read_csv('/Users/laxmigupte/Desktop/Ironhack/week4/D3/lab-hypothesis-testing-en/data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [3]:
# Your code here
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


In [4]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [5]:
salaries.dtypes

Name                  object
Job Titles            object
Department            object
Full or Part-Time     object
Salary or Hourly      object
Typical Hours        float64
Annual Salary        float64
Hourly Rate          float64
dtype: object

In [6]:
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [24]:
# Your code here
from scipy.stats import ttest_1samp
import scipy.stats as st

In [29]:
hr = salaries[salaries['Salary or Hourly']=='Hourly']['Hourly Rate'].dropna()
hr

4        19.86
6        46.10
7        35.60
10        2.65
18       17.68
         ...  
33164    46.10
33168    17.68
33169    35.60
33174    46.35
33175    48.85
Name: Hourly Rate, Length: 8022, dtype: float64

In [30]:
samp1 = hr
print("samp1 mean: ", samp1.mean())

samp1 mean:  32.78855771628024


In [31]:
test = st.ttest_1samp(hr,30)

In [32]:
test.pvalue

4.3230240486229894e-92

In [33]:
type(test)

scipy.stats._stats_py.TtestResult

In [37]:
test.pvalue<0.05

True

In [38]:
alpha = 0.05
if test.pvalue < alpha:
    print("Reject the Null Hypothesis: The wage is significantly different from $30.")
else:
    print("Fail to Reject the Null Hypothesis.")

Reject the Null Hypothesis: The wage is significantly different from $30.



The average hourly wage ($32.78$) is significantly different from $30/hr.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [45]:
# Your code here
sample_mean = hr.mean()
sample_mean

32.78855771628024

In [46]:
sample_size = len(hr)
sample_size

8022

In [47]:
#Degrees of Freedom: Sample size minus 1.
degrees_freedom = sample_size - 1
degrees_freedom

8021

In [48]:
hr.describe()

count    8022.000000
mean       32.788558
std        12.112573
min         2.650000
25%        21.200000
50%        35.600000
75%        40.200000
max       109.000000
Name: Hourly Rate, dtype: float64

In [51]:
# Standard Error = Standard Deviation / sqrt(n)
standard_error = st.sem(hr)
standard_error

0.1352368565101596

In [52]:
con_int = st.t.interval(confidence=0.95, 
                             df=degrees_freedom, 
                             loc=sample_mean, 
                             scale=standard_error)



In [53]:
print(f"95% Confidence Interval: {con_int}")

95% Confidence Interval: (32.52345834488425, 33.05365708767623)


We are 95% confident that the actual average hourly wage for all workers in the city is between 32.52% and 33.05%

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [54]:
# Your code here
# Number of hourly workers
count = salaries[salaries['Salary or Hourly'] == 'Hourly'].shape[0]

# Total number of workers
nobs = salaries.shape[0]

# The proportion we are testing against (25%)
value = 0.25

In [55]:
from statsmodels.stats.proportion import proportions_ztest

In [56]:
z_stat, p_value = proportions_ztest(count=count, nobs=nobs, value=value)

print(f"Z-statistic: {z_stat}")
print(f"P-value: {p_value}")

Z-statistic: -3.5099964213703005
P-value: 0.0004481127249057967


In [57]:
alpha = 0.05
if test.pvalue < alpha:
    print("Reject the Null Hypothesis: The proportion of hourly workers is significantly different from 25%.")
else:
    print("We don't have enough evidence to say it's different from 25%.")

Reject the Null Hypothesis: The proportion of hourly workers is significantly different from 25%.
